Add guidelines for code review (#134)
bemanproject/beman: guidelines/code-review.md

<!--
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

# Code Review Guidelines

The following documents code review processes for the Beman Project organization.

* As stated in the Beman Standard, [REPOSITORY.CODEOWNERS](https://github.com/bemanproject/beman/blob/main/docs/BEMAN_STANDARD.md#repositorycodeowners) is a REQUIREMENT, thus a list of local reviewers is available for each repository.

* By default, all PRs are reviewed by at least one code owner.

> * PRs created by codeowners can be merged without review. RECOMMENDATION: A 24-hour grace period is provided for other code owners to review the PR.

Reviewer: ```suggestion
* PRs created by codeowners can be merged without review. 
 
RECOMMENDATION: A 24-hour grace period is provided for other code owners to review the PR.
```

<!--
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

# Code Review Guidelines

The following documents code review processes for the Beman Project organization.

* As stated in the Beman Standard, [REPOSITORY.CODEOWNERS](https://github.com/bemanproject/beman/blob/main/docs/BEMAN_STANDARD.md#repositorycodeowners) is a REQUIREMENT, thus a list of local reviewers is available for each repository.

> 

Author: ```suggestion
```

<!--
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

# Code Review Guidelines

> Short summary of the code review process in the Beman Project organization.

Reviewer 2: ```suggestion
The following documents code review processes for the Beman Project organization.
```

> * As stated in the Beman Standard, [REPOSITORY.CODEOWNERS](https://github.com/bemanproject/beman/blob/main/docs/BEMAN_STANDARD.md#repositorycodeowners) is a REQUIREMENT, thus a list of local reviewers is available for each repository.

Reviewer 3: A little more friendly...

```suggestion
* As stated in [REPOSITORY.CODEOWNERS](https://github.com/bemanproject/beman/blob/main/docs/BEMAN_STANDARD.md#repositorycodeowners), a list of local reviewers is available for each repository.
```

* By default, all PRs are reviewed by at least one code owner.

> * PRs created by codeowners can be merged without review. 
>  
> RECOMMENDATION: A 24-hour grace period is provided for other code owners to review the PR.

Reviewer 3: ```suggestion
* PRs created by codeowners can be merged without review, but it is a good practice to offer other codeowners an opportunity to review the PR by providing a 24-hour grace period.
```

> > If you find a repository without a CODEOWNERS file or without a rule set (e.g., PR and code review enforced before merging), please report to Beman Leads or on [Discourse](https://discourse.bemanproject.org) or raise a PR to fix the issue.

Author: ```suggestion
```

> > If you find a repository without a CODEOWNERS file or without a rule set (e.g., PR and code review enforced before merging), please report to Beman Leads or on [Discourse](https://discourse.bemanproject.org) or raise a PR to fix the issue.

Author: Done

> * Internal contributors also require a review from a code owner. As an exception, if no review is received within 24 hours, the code owner can merge the PR, but notify the reviewers. Any reported issues will be fixed by the same code owner.

Reviewer 2: ```suggestion
* Internal contributions also require a review from a code owner. However, if no review is received within 24 hours, the code owner can merge the PR. If issues arise the responsible code owner should resolve.
```

>   - Note: Internal contributors require to have admin rights to bypass the review process in exception cases. Please ping Beman Leads or on [Discourse](https://discourse.bemanproject.org) to request admin rights ahead.

Reviewer 2: I mean if they are a repo owner I think they have these rights? I'm not sure this note adds anything?

Author: That's not true. It depends on 1) who created the repo (e.g., it was created by X on personal account and then moved or it was created by an org admin and 2) what are the rules set in that repo (by default, admin cannot bypass rules, but they can be configured so):

![image](https://github.com/user-attachments/assets/4cf84595-f7e9-4732-8452-bed01c5c92b2)

Reviewer 2: that's fair -- some suggested wordsmitting...

* External contributors require a review from a code owner. If the tagged owner is not available, please ping another one or request a review on [Discourse](https://discourse.bemanproject.org).

* Internal contributions also require a review from a code owner. However, if no review is received within 24 hours, the code owner can merge the PR. If issues arise the responsible code owner should resolve.

>   - Note: Internal contributors require to have admin rights to bypass the review process in exception cases. Please ping Beman Leads or on [Discourse](https://discourse.bemanproject.org) to request admin rights ahead.

Reviewer 2: ```suggestion
  - Note: Internal contributors must have admin rights to bypass the review process in exception cases. Please ping Beman Leads or on [Discourse](https://discourse.bemanproject.org) to request admin rights.
```
